File: SJF.R
cat("\014")

arrivals <- c(0, 3, 4, 8, 12)
ext <- c(15, 2, 6, 2, 10)

if (length(arrivals) == length(ext)) {
  cat("Number of Process:", length(arrivals), "\n")
  cat("Total Burst Time:", sum(ext), "\n")
  
  n <- length(arrivals)
  process <- data.frame(
    Process = paste0("P", 1:n),
    AT = arrivals,
    BT = ext,
    ST = numeric(n),
    CT = numeric(n),
    WT = numeric(n),
    TT = numeric(n),
    Completed = logical(n)
  )
  
  # SJF (Non-preemptive) Scheduling Algorithm
  current_time <- 0
  completed_count <- 0
  
  while (completed_count < n) {
    available <- which(process$AT <= current_time & !process$Completed)
    
    if (length(available) > 0) {
      selected <- available[which.min(process$BT[available])]
      if (length(selected) > 1) {
        selected <- selected[which.min(process$AT[selected])]
      }
      
      process$ST[selected] <- current_time
      process$CT[selected] <- current_time + process$BT[selected]
      process$TT[selected] <- process$CT[selected] - process$AT[selected]
      process$WT[selected] <- process$TT[selected] - process$BT[selected]
      process$Completed[selected] <- TRUE
      current_time <- process$CT[selected]
      completed_count <- completed_count + 1
    } else {
      next_arrival <- min(process$AT[!process$Completed])
      current_time <- next_arrival
    }
  }
  
  cat("\n")
  cat("SJF\tP\tAT\tBT\tST\tCT\tWT\tTT\n")
  for (i in 1:n) {
    cat(sprintf("\t%s\t%d\t%d\t%d\t%d\t%d\t%d\n", 
                process$Process[i], 
                process$AT[i], 
                process$BT[i], 
                process$ST[i], 
                process$CT[i], 
                process$WT[i], 
                process$TT[i]))
  }
  cat("Average WT:", mean(process$WT), "\n")
  cat("Average TT:", mean(process$TT), "\n")
} else {
  stop("Arrays have different lengths.")
}

File: SJF.py
import numpy as np
import pandas as pd

ARRIVALS = np.array([0, 3, 4, 8, 12])
BURST = np.array([15, 2, 6, 2, 10])

def sjf(ARRIVALS, BURST):
    n = len(ARRIVALS)
    ST = np.zeros(n, dtype=int)  # Initialize Start Time
    CT = np.zeros(n, dtype=int)  # Initialize Completion Time
    WT = np.zeros(n, dtype=int)  # Initialize Waiting Time
    TT = np.zeros(n, dtype=int)  # Initialize Turnaround Time
    completed = np.zeros(n, dtype=bool)  # Track completed processes
    remaining_burst = BURST.copy()
    
    current_time = 0
    completed_count = 0
    
    while completed_count < n:
        shortest_job = -1
        shortest_burst = float('inf')
        
        for i in range(n):
            if not completed[i] and ARRIVALS[i] <= current_time:
                if remaining_burst[i] < shortest_burst:
                    shortest_burst = remaining_burst[i]
                    shortest_job = i
        
        if shortest_job == -1:
            next_arrival = min(ARRIVALS[i] for i in range(n) if not completed[i])
            current_time = next_arrival
            continue
        
        ST[shortest_job] = current_time
        CT[shortest_job] = current_time + BURST[shortest_job]
        WT[shortest_job] = ST[shortest_job] - ARRIVALS[shortest_job]
        TT[shortest_job] = CT[shortest_job] - ARRIVALS[shortest_job]
        
        current_time = CT[shortest_job]
        completed[shortest_job] = True
        completed_count += 1
    
    df = pd.DataFrame({
        'Process': [f'P{i+1}' for i in range(n)],
        'AT': ARRIVALS,
        'BT': BURST,
        'ST': ST,
        'CT': CT,
        'WT': WT,
        'TT': TT
    })
    
    avg_waiting = np.mean(WT)
    avg_turnaround = np.mean(TT)
    
    print(f"\nShortest Job First (SJF) Scheduling - Non-Preemptive\n")
    print(df.to_string(index=False))
    print(f"Average Waiting Time: {avg_waiting:.2f}")
    print(f"Average Turnaround Time: {avg_turnaround:.2f}")
    
    return df, avg_waiting, avg_turnaround

sjf(ARRIVALS, BURST)